### ISSUE TITLE
Large data model for built_value causes large memory consumption and codegen runtime

### ISSUE THREAD

**Issue by michael-golfi** [dart-lang/build#3281]
Hello, @Markzipan is a friend of mine and encouraged me to post this issue I'm seeing with build_runner and built_value.

I'm running the following tooling versions:

* Dart SDK version: 2.16.2 (stable) (Tue Mar 22 13:15:13 2022 +0100) on "macos_x64"
* build_runner: 2.1.8
* built_value: 8.1.4

### Problem

Our data model is generated from a GraphQL schema and operations which describe the data being pulled from the API. The data model has quite a few references between built values within the output serializable types. These are the following packages we're using for covering that portion of the builder process.

* Other packages with versions for builders:
* gql_build: 0.4.1
* gql_code_builder: 0.5.1
* ferry_generator: 0.5.0-dev.10

Though this doesn't seem relevant to the issue, I am building this for Flutter on iOS and Android.

### My investigation

I started investigating our issues with codegen once I noticed that codegen started going up to roughly 22 mins on average and previously also went higher (eg: 1 hour). At the same time I'm seeing memory consumption of 22 Gb of RAM on a Macbook Pro (it only has 16 Gb of physical RAM). It seems like i/o also goes high and some of the files take several minutes to generate output from `built_value`. I started seeing messages such as the following:

```text
[WARNING] No actions completed for 15.0s, waiting on:
  - built_value_generator:built_value on lib/src/graphql/resolvers/__generated__/login.queries.data.gql.dart
  - built_value_generator:built_value on lib/src/graphql/resolvers/__generated__/login.queries.var.gql.dart
  - built_value_generator:built_value on lib/src/graphql/resolvers/__generated__/login.queries.req.gql.dart
  - built_value_generator:built_value on lib/src/graphql/resolvers/__generated__/delete_event.mutations.ast.gql.dart
  - built_value_generator:built_value on lib/src/graphql/resolvers/__generated__/delete_event.mutations.data.gql.dart
  .. and 11 more

...

[WARNING] No actions completed for 7m 54s, waiting on:
  - built_value_generator:built_value on lib/src/graphql/resolvers/__generated__/event_tracking.mutations.data.gql.dart
  .. and 11 more
```

I ran `--track-performance` to see the timing graph and noticed high latency for a few generation tasks (the same ones which are mentioned in the `[WARNING]` message above. More and more it started like there's some form of blocking / resource contention occurring due to waiting on certain operations to finish for others to proceed.

<img width="2219" alt="Screen Shot 2022-04-11 at 12 53 27 PM" src="https://user-images.githubusercontent.com/1388500/162793986-2689a8a4-f2e5-4668-b9f4-f6a01dc6b7f6.png">

### What I've tried

I read through a few issues which seem to suggest that build runner doesn't do well with parallel operations on the asset graph (eg: single process access if my memory serves right?) and no shared memory access. I read [this comment](https://github.com/dart-lang/build/issues/2583#issuecomment-568549837) and decided to set `BUILD_MAX_WORKERS_PER_TASK` and ran it a few times between `[1, 4]` to get an idea of what the throughput limitations are.

I found that `BUILD_MAX_WORKERS_PER_TASK` values of 1 and 2 seem to be reasonable. I also split some of the files to make them smaller, which also makes the intermediate files (output of GraphQL operations in Dart with JSON serializer annotations, input to built_value) smaller. Reducing the size of a particular couple of files made the codegen ~4x faster. Setting that environment variable seemed to limit the amount of RAM used during the process.

I wanted to bring this to your attention and hope my own debugging helps improve the tool. Happy to share some more if necessary.

**Comment by jakemac53** [comment]
Have you tried using `generate_for` config in order to ensure that `built_value` is only running on the builders you want it to run on? See the docs here https://github.com/dart-lang/build/blob/master/docs/faq.md#how-do-i-avoid-running-builders-on-unnecessary-inputs

Generally these very slow builds happen on large projects, where the builders are running on all dart files in the entire project (which is the default), instead of just on the files that they need to actually be ran on. This results in both slow execution but also dramatically increased memory usage (both become worst case O(n^2) based on the number of source files).

This is particularly true if your entire app is a single large package (which is common, and not "wrong", but does cause some performance issues with build_runner).

Another thing you can do to really dramatically help performance is to keep the libraries that have codegen ran on them as small as possible. Each of these libraries in the build ends up with a dependency on all the transitive libraries that it imports (which is where the worst case n^2 behavior comes from).

tldr; Keep your model libraries as small as possible, and configure your builders to only run on those libraries

**Comment by michael-golfi** [comment]
Yes, I suppose I unintentionally left out that detail. I tried in a bare package with minimal dependencies and Ferry's most recent update also seems to glob only the necessary files for input into `built_value`.

In the end I managed to get codegen down to 4.5 minutes. Ferry runs in ~10-15 seconds (consistently) for creating the input files to `built_value`. I didn't encounter a significant speedup due to separating into its own project. I saw the major speedup by reorganizing the input GraphQL operations into individual files.

**Comment by jakemac53** [comment]
Note that the [built_value generator](https://github.com/google/built_value.dart/blob/master/built_value_generator/build.yaml#L12) is configured to run on _all_ dart files. So if you aren't configuring it explicitly to run on fewer, it likely is part of the issue.

**Comment by jakemac53** [comment]
Specifically for instance it looks like you could try doing something like the following:

```yaml
targets:
  $default:
    builders:
      built_value_generator:built_value:
        generate_for:
          - **/*.gql.dart
```

**Comment by michael-golfi** [comment]
Thanks, I can also try this. I ran build_runner in the package template created by `flutter create -t package`, which I think only has one dart file.

**Comment by jakemac53** [comment]
> Thanks, I can also try this. I ran build_runner in the package template created by `flutter create -t package`, which I think only has one dart file.

Are you saying you were experiencing these long build times on that small app as well? Not quite understanding what you mean by that :).

**Comment by michael-golfi** [comment]
Yes, sorry for my brevity, I will try to elaborate as much as I can on that. I will run this again on my original GraphQL files in an empty project:

Now I have moved all of our GraphQL files from our flutter app directory with the following layout and respective sizes:

```text
├── __generated__/
|   └── Generated schema files: *.gql.dart, *.gql.g.dart
├── main.dart
├── schema.graphql
└── src
    ├── features
    |   └── directories and dart files for our app
    ├── graphql
    │   ├── resolvers
    │   │   ├── __generated__/
    │   │   |   └── *.gql.dart, *.gql.g.dart
    │   │   ├── fragments
    │   │   │   ├── __generated__/
    │   │   │   |  └── *.gql.dart, *.gql.g.dart
    │   │   │   └── all graphql fragment files: *.graphql
    │   │   └── all graphql operation files: *.graphql
    │   └── serializers
    |       └── all custom serializers: *.dart
    └── widgets/
        └── ... some flutter widgets...

88 directories, 738 files
```

into an empty project created by `flutter create -t package`. It only contains 60 files, of which they are only the GraphQL files and 5 custom dart serializer files, all of which are inputted to the codegen. Now I'll run `build_runner` on this. 

```text
lib/
├── schema.graphql
└── src
    └── graphql
        ├── resolvers
        │   ├── fragments
        │   │   └── *_fragment.graphql
        │   └── *.graphql
        └── serializers
            └── *.dart

5 directories, 60 files
```

Running it in the empty project yields this (this is approximately where it maxed out):

<img width="742" alt="Screen Shot 2022-04-11 at 4 36 15 PM" src="https://user-images.githubusercontent.com/1388500/162827763-1e723f09-3e57-47d2-ba74-6e3cec4c0450.png">

At this point it's usually stalled out and emits the timeout warning:

```text
[INFO] 2m 0s elapsed, 261/277 actions completed.
[INFO] 2m 2s elapsed, 265/279 actions completed.
[INFO] 2m 4s elapsed, 269/281 actions completed.
[INFO] 2m 7s elapsed, 273/281 actions completed.
[INFO] 8m 43s elapsed, 277/281 actions completed.
[WARNING] No actions completed for 6m 36s, waiting on:
  - built_value_generator:built_value on lib/src/graphql/resolvers/__generated__/event_tracking.mutations.data.gql.dart
  - built_value_generator:built_value on lib/src/graphql/resolvers/__generated__/event_tracking.mutations.var.gql.dart
  - built_value_generator:built_value on lib/src/graphql/resolvers/__generated__/event_tracking.mutations.req.gql.dart
  - built_value_generator:built_value on lib/__generated__/serializers.gql.dart

[INFO] 8m 48s elapsed, 277/281 actions completed.
[INFO] 11m 55s elapsed, 279/281 actions completed.
[WARNING] No actions completed for 3m 7s, waiting on:
  - built_value_generator:built_value on lib/src/graphql/resolvers/__generated__/event_tracking.mutations.req.gql.dart
  - built_value_generator:built_value on lib/__generated__/serializers.gql.dart

[INFO] 11m 56s elapsed, 323/339 actions completed.
[INFO] 11m 57s elapsed, 364/375 actions completed.
[INFO] 11m 59s elapsed, 404/411 actions completed.
[INFO] 12m 0s elapsed, 445/450 actions completed.
[INFO] 12m 1s elapsed, 474/490 actions completed.
[INFO] 12m 2s elapsed, 505/505 actions completed.
[INFO] Running build completed, took 12m 2s

[INFO] Caching finalized dependency graph...
[INFO] Caching finalized dependency graph completed, took 628ms

[INFO] Succeeded after 12m 3s with 548 outputs (3645 actions)
```

Just for my own sanity I'll run it on the full project once again and post the results as soon as `build_runner` finishes. Although 12 minutes looks faster, I have been seeing 22 minutes in the bigger project.

**Comment by michael-golfi** [comment]
Running it in our actual repo with all of the other dart files seems to complete at around the same mark at 11m58s. Though this is just one run against one run for either case. Not exactly a great benchmark... There do seem to be some other significant metrics.

I still saw some stalling in the log:

<details>

```text
[INFO] 2m 1s elapsed, 335/350 actions completed.
[INFO] 2m 3s elapsed, 336/351 actions completed.
[INFO] 2m 5s elapsed, 337/352 actions completed.
[WARNING] No actions completed for 5m 57s, waiting on:
  - built_value_generator:built_value on lib/src/graphql/resolvers/event_tracking.mutations.data.gql.dart
  - built_value_generator:built_value on lib/schema.schema.gql.dart
  .. and 10 more

[INFO] 8m 9s elapsed, 339/354 actions completed.
[INFO] 8m 11s elapsed, 378/392 actions completed.
[INFO] 8m 17s elapsed, 392/395 actions completed.
[WARNING] No actions completed for 2m 49s, waiting on:
  - built_value_generator:built_value on lib/serializers.gql.dart

[INFO] 11m 8s elapsed, 395/411 actions completed.
[INFO] 11m 9s elapsed, 450/466 actions completed.
[INFO] 11m 12s elapsed, 455/471 actions completed.
[INFO] 11m 13s elapsed, 461/476 actions completed.
[INFO] 11m 14s elapsed, 471/486 actions completed.
[INFO] 11m 15s elapsed, 479/495 actions completed.
[INFO] 11m 16s elapsed, 487/503 actions completed.
[INFO] 11m 17s elapsed, 492/508 actions completed.
[INFO] 11m 18s elapsed, 501/517 actions completed.
[INFO] 11m 19s elapsed, 508/524 actions completed.
[INFO] 11m 20s elapsed, 518/534 actions completed.
[INFO] 11m 21s elapsed, 526/542 actions completed.
[INFO] 11m 22s elapsed, 536/551 actions completed.
[INFO] 11m 23s elapsed, 548/561 actions completed.
[INFO] 11m 25s elapsed, 555/571 actions completed.
[INFO] 11m 26s elapsed, 566/582 actions completed.
[INFO] 11m 27s elapsed, 576/592 actions completed.
[INFO] 11m 28s elapsed, 591/605 actions completed.
[INFO] 11m 29s elapsed, 600/616 actions completed.
[INFO] 11m 30s elapsed, 610/626 actions completed.
[INFO] 11m 31s elapsed, 620/635 actions completed.
[INFO] 11m 32s elapsed, 629/645 actions completed.
[INFO] 11m 33s elapsed, 643/659 actions completed.
[INFO] 11m 34s elapsed, 664/678 actions completed.
[INFO] 11m 35s elapsed, 680/696 actions completed.
[INFO] 11m 36s elapsed, 696/712 actions completed.
[INFO] 11m 37s elapsed, 706/721 actions completed.
[INFO] 11m 39s elapsed, 715/731 actions completed.
[INFO] 11m 40s elapsed, 724/740 actions completed.
[INFO] 11m 41s elapsed, 734/749 actions completed.
[INFO] 11m 42s elapsed, 764/780 actions completed.
[INFO] 11m 43s elapsed, 802/817 actions completed.
[INFO] 11m 44s elapsed, 825/840 actions completed.
[INFO] 11m 45s elapsed, 847/862 actions completed.
[INFO] 11m 46s elapsed, 858/873 actions completed.
[INFO] 11m 47s elapsed, 899/914 actions completed.
[INFO] 11m 49s elapsed, 916/932 actions completed.
[INFO] 11m 50s elapsed, 917/932 actions completed.
[INFO] 11m 51s elapsed, 979/995 actions completed.
[INFO] 11m 52s elapsed, 1011/1022 actions completed.
[INFO] 11m 53s elapsed, 1033/1046 actions completed.
[INFO] 11m 54s elapsed, 1061/1068 actions completed.
[INFO] 11m 55s elapsed, 1080/1094 actions completed.
[INFO] 11m 56s elapsed, 1105/1121 actions completed.
[INFO] Running build completed, took 11m 57s

[INFO] Caching finalized dependency graph...
[INFO] Caching finalized dependency graph completed, took 517ms

[INFO] Succeeded after 11m 58s with 547 outputs (1362 actions)
```

</details>

I noticed there also seems to be a lot of read ops:

<img width="737" alt="Screen Shot 2022-04-11 at 5 00 40 PM" src="https://user-images.githubusercontent.com/1388500/162831319-beffcefa-d45a-4d08-be9e-092947ce09f3.png">

And it used slightly more memory:

<img width="734" alt="Screen Shot 2022-04-11 at 4 58 01 PM" src="https://user-images.githubusercontent.com/1388500/162831483-48d9bfa5-260b-492e-af17-ec4484397c06.png">

**Comment by michael-golfi** [comment]
I can share the empty project privately, if you wouldn't mind? Our schema isn't exactly public access.

**Comment by jakemac53** [comment]
Unfortunately I can't look at any private code, would it be possible to make a public project with a fake schema that exhibits the same behavior?

**Comment by jakemac53** [comment]
This definitely does work to repro the issue, thanks.

**Comment by jakemac53** [comment]
I don't know exactly how this ferry_generator package works, or what its constraints might be, but it seems like it is probably trying to aggregate all the schemas into a single large library.

This will cause very poor incremental build performance as well, as you have seen. If you touch any schema it will cause that large library to be re-generated, built_value to have to re-run on the entire schema, etc.

I would suggest to the authors of `ferry_generator` that they possibly explore other strategies here. For instance, can the large `schema.schema.gql.dart` files actually just contain exports of several other, smaller libraries (possibly generated next to the other gql files for each schema?).

Those individual libraries could be incrementally generated more quickly, and not so much work would have to be redone on any change.

**Comment by jakemac53** [comment]
I added some more time tracking in for the resolver, which I will actually upstream as I think it is pretty useful, but it does in any case confirm that just the call to `libraryFor` on the schema dart file takes about 40 seconds.

This is all inside of the `getLibraryForUri` method from the analyzer.

One interesting thing I noted from that, is it reports the `user` time as 0 seconds. I think this is probably something to do with the implementation of `getLibraryForUri` though, our tracking is not perfect and relies on zones. I assume the real work here is happening in some callback not scheduled by our call to `getLibraryForUri`, and it is just waiting for some future to be completed or something, from a task scheduled by a different zone.

**Comment by michael-golfi** [comment]
Thanks @jakemac53 for this detailed response, really appreciate it and I hope this has all been helpful for you all for performance data. Typically the schema files are one single file which represents the "global data model" of a GraphQL API. They're typically the response of an introspection query from a GraphQL API.

I think there would need to be quite a bit of thought for splitting it. Would that splitting need to minimize inter-file dependencies or where types which are related need to be grouped together?

**Comment by jakemac53** [comment]
I am in any case going to close this as working as expected today, I don't think there is really anything we can do in particular here. These are just very large generated libraries, which take a very long time to analyze.

The ferry_generator should probably consider not relying on a separate `built_value` codegen step, and directly generate all code from the graphql files.

**Comment by jakemac53** [comment]
> I think there would need to be quite a bit of thought for splitting it. Would that splitting need to minimize inter-file dependencies or where types which are related need to be grouped together?

You would want to minimize inter-file dependencies as much as possible, yes. But having some cycles shouldn't be any worse than grouping the files together.
